### PR TITLE
Fixed #29522 -- Refactored the Deserializer functions to classes.

### PR DESCRIPTION
**Pull request by AMK9978**
# Trac ticket number

ticket-29522

# Branch description
Updated the [previous PR](https://github.com/django/django/pull/11218).
I applied the latest comments and also added a test to show that a developer can extend the Deserializer class and override its methods, although there is not a lot to be overridden. Extensibility is its main point.

# Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: docs/topics/serialization.txt
Comment (AMK9978) [inline docs/topics/serialization.txt:151]: I know this may be lame but waiting for your comments. The same for the release note.

File: django/core/serializers/python.py
Comment (sarahboyce) [inline django/core/serializers/python.py:99]: We already have `django.core.serializers.base.Deserializer` which the xml deserializer inherits
Then `json`, `jsonl` and `pyaml` each have `Deserializer` functions still, it would be great if this could be adapted so that these are all classes that inherit from `python.Deserializer` and the python `Deserializer` inherits from `base.Deserializer`
The current state is a bit messy which I think makes it harder to document :thinking:

File: django/core/serializers/python.py
Comment (sarahboyce) [inline django/core/serializers/python.py:208]: If we look at `xml.Deserializer` this has `_get_model_from_node`, `_handle_m2m_field_node`, `_handle_fk_field_node` and `__next__` instead of `__iter__` is defined - it would be good to align this

**Review by sarahboyce — COMMENTED** [review]
Thank you for picking this up and being patient @AMK9978  :star: 
I have a few thoughts on the design here :+1:

File: django/core/serializers/python.py
Comment (AMK9978) [reply django/core/serializers/python.py:99]: Fair enough. So, like what happens in the `xml_serializer`, creating the object list in `json`, `jsonl`, and `pyaml` which are now created using their `Deserializer` helper functions in `django/core/serializers/json.py`, `django/core/serializers/jsonl.py`, and `django/core/serializers/pyaml.py`, respectively, should become a method in the `Deserializer` class (the equivalent of `pulldom.parse(self.stream, self._make_parser())` in the `xml_deserializer`). For each three of them, there must be a class to implement that object creation🤔

**Comment by AMK9978** [comment]
@sarahboyce 
I tried to make the deserializer classes look more similar with the least changes and still readable code. Waiting for your comments!

File: django/core/serializers/python.py
Comment (sarahboyce) [inline django/core/serializers/python.py:124]: This is untested/missing coverage

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @AMK9978 :star: this is shaping up
I have a few more suggestions/comments

**Comment by AMK9978** [comment]
@sarahboyce Thanks for your comments. While I was keeping the classes alike, your simplification removed boilerplates. Loved it!
Btw, your suggestions needed some tweaks. For example, possible errors from `_handle_object` should be handled by the subclasses as they used to and tests require that. Plus, I added some tests for the functionality of the deserializer in Django.
Waiting for the next round of your comments :) (or approval!).

File: django/core/serializers/jsonl.py
Comment (sarahboyce) [inline django/core/serializers/jsonl.py:58]: I can remove this `_handle_object` without any test failures

File: django/core/serializers/jsonl.py
Comment (sarahboyce) [inline django/core/serializers/jsonl.py:64]: This appears not to be covered anymore, maybe due to the change to `.splitlines()` :thinking:

File: django/core/serializers/pyyaml.py
Comment (sarahboyce) [inline django/core/serializers/pyyaml.py:87]: I can remove this without any test failures

File: django/core/serializers/jsonl.py
Comment (sarahboyce) [inline django/core/serializers/jsonl.py:48]:
```suggestion
        if isinstance(stream_or_string, (bytes, str)):
```
This was also ran for bytes previously, was this a conscious change? Might be worth adding a test for bytes input (there isn't one in the existing test suite) to check everything is ok or this makes sense
The pyyaml also is missing coverage for bytes input
Both of these are existing coverage issues but it would be nice to resolve them considering we refactoring this area

File: docs/topics/serialization.txt
Comment (sarahboyce) [reply docs/topics/serialization.txt:151]: I think maybe we should mention how folks can add their own serialization format, so for example we should mention the `SERIALIZATION_MODULES` setting and how they can extend extisting `Deserializer`s

I think this should be within the `Serialization formats` section, maybe at the bottom after `YAML`, as `Custom  serialization formats` (or something). This can have a `.. _custom-serialization-formats:` reference link we can use for the release notes.

It will also need a `.. versionchanged: 5.2` note

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @AMK9978 :star:
I think this is shaping up nicely :1st_place_medal:

File: django/core/serializers/jsonl.py
Comment (AMK9978) [reply django/core/serializers/jsonl.py:58]: Because of insufficient tests. For example, if you remove `_handle_object` in the json.py and run `python runtests.py fixtures_regress`, you'll get ` Animal has no field named 'extra_name'`. I will add corresponding tests for these scenartios for `jsonl.py` and `pyyaml.py`.

File: django/core/serializers/jsonl.py
Comment (AMK9978) [reply django/core/serializers/jsonl.py:48]: I didn't quite get what you meant because that check for bytes was unnecessary. If there's a `bytes` object, it gets converted into `str` in its above `if`. Anyway, I have added tests for bytes input and the code works well.

File: django/core/serializers/jsonl.py
Comment (AMK9978) [reply django/core/serializers/jsonl.py:64]: I don't think changing `split('\n')` to `splintlines` changed anything else but I added a test for a jsonl input with empty lines.

File: docs/topics/serialization.txt
Comment (sarahboyce) [reply docs/topics/serialization.txt]: I have tried to rewrite the docs a bit, can you review the updates? Would be good to get more eyes on them for sure

**Review by sarahboyce — COMMENTED** [review]
Thank you for the tests @AMK9978 :star: this looks really good 

I have squashed and tweaked slightly, I would like another pair of eyes on the docs and then I think we're good to go :+1:

File: docs/topics/serialization.txt
Comment (AMK9978) [inline docs/topics/serialization.txt:366]: @sarahboyce 
Awesome job! 
IMHO, I don't think custom deserialization quite fits into the  `Custom serialization formats`. Plus, adding a custom serializer to the `SERIALIZATION_MODULES` is a convenient way that only applies to serializers but not such a thing for a custom deserializer. So it doesn't seem cohesive to me.

Comment (sarahboyce) [reply docs/topics/serialization.txt:366]: I think this is how folks are supposed to do it? `get_deserializer` loads the serializers defined by this setting, and then gets the `Deserializer` from it. So this is how you define both a custom serializer and deserializer

**Comment by pauloxnet** [comment]
How Django performances change with this change?

File: docs/topics/serialization.txt
Comment (AMK9978) [reply docs/topics/serialization.txt:366]: Yes, you're right about the usage and how `get_deserializer` returns and I was confused. I wanted to suggest a direct mention of custom deserialization in the section title. Plus, regarding 
`First, define a ``Serializer`` and a ``Deserializer`` class. `
having a custom Deserialization doesn't raise any error in the custom module unless deserialization for that custom format gets called. Maybe it indicates it's necessary (Not a big deal!).

Comment (sarahboyce) [reply docs/topics/serialization.txt:366]: > having a custom Deserialization doesn't raise any error in the custom module unless deserialization for that custom format gets called. Maybe it indicates it's necessary (Not a big deal!).

I'm not quite sure what the suggestion is, do you think this needs a docs warning? Feel free to push up updates :+1:

Comment (AMK9978) [reply docs/topics/serialization.txt:366]: Got it. I'll wait for other reviewers' comments. If the docs get no comment, it means it's clear enough.

**Comment by sarahboyce** [comment]
> How Django performances change with this PR?

- Ran the benchmarks and the output is "no significant change" ( see [here](https://github.com/django/django/actions/runs/10848956962/))
- Ran some tests on main vs this branch and don't see a significant impact to performance

Is there a specific area that concerns you @pauloxnet ?

Whilst we have you, can you check if you're happy with the docs?

File: docs/topics/serialization.txt
Comment (pauloxnet) [inline docs/topics/serialization.txt:355]:
```suggestion
In addition to the default formats, you can create a custom serialization 
format.
```
I tried to rephrase, but feel free to discard it.

File: docs/topics/serialization.txt
Comment (pauloxnet) [reply docs/topics/serialization.txt:366]: I think we should put an example of Serializer and Deserializer with minimal but working code, and not empty.

File: docs/topics/serialization.txt
Comment (pauloxnet) [inline docs/topics/serialization.txt:372]: This path here might be confusing if it is not defined above what it is. Maybe we could define above that the code goes in a python file with a certain name and then call it here to help understand that this is the path to the python file (or module) that contains the serializers.

**Comment by pauloxnet** [comment]
> > How Django performances change with this PR?
> 
>     * Ran the benchmarks and the output is "no significant change" ( see [here](https://github.com/django/django/actions/runs/10848956962/))
> 
>     * Ran some tests on main vs this branch and don't see a significant impact to performance
> 
> 
> Is there a specific area that concerns you @pauloxnet ?

nothing in particular, I just thought it was a good idea to verify that switching from functions to classes didn't degrade performance.

> Whilst we have you, can you check if you're happy with the docs?

Overall it all seems ok to me, I just tried to give some suggestions, but feel free to discard them if they don't make sense to you.

File: docs/topics/serialization.txt
Comment (sarahboyce) [reply docs/topics/serialization.txt:372]: Yes! @AMK9978 something like
```
.. code-block:: python
    :caption: ``path/to/custom_serializer.py``
```
should work

File: docs/topics/serialization.txt
Comment (sarahboyce) [reply docs/topics/serialization.txt:366]: Yeah I agree, a more realistic example would be beneficial. @AMK9978 do you mind pulling something together?

File: docs/topics/serialization.txt
Comment (pauloxnet) [reply docs/topics/serialization.txt:372]: Yes, exactly.

File: docs/topics/serialization.txt
Comment (AMK9978) [reply docs/topics/serialization.txt]: @pauloxnet @sarahboyce 
I added a minimal functioning csv serializer and deserializer. Although due to the flat nature of csv, it requires a bit more code, I think it is quite useful as an example.

File: docs/topics/serialization.txt
Comment (sarahboyce) [inline docs/topics/serialization.txt:407]: I haven't tested but I am hoping we can do something like...
```suggestion
    class Serializer(serializers.python.Serializer):
        def get_dump_object(self, obj):
            dumped_object = super().get_dump_object(obj)
            row = [dumped_object["model"], str(dumped_object["pk"])]
            row += [str(value) for value in dumped_object["fields"].values()]
            return ",".join(row), dumped_object["model"]

        def end_object(self, obj):
            dumped_object_str, model = self.get_dump_object(obj)
            if self.first:
                fields = [field.name for field in apps.get_model(model)._meta.fields]
                header = ",".join(fields)
                self.stream.write(f"model,{header}\n")
            self.stream.write(f"{dumped_object_str}\n")

        def getvalue(self):
            return super(serializers.python.Serializer, self).getvalue()
```

File: docs/topics/serialization.txt
Comment (sarahboyce) [inline docs/topics/serialization.txt:430]: Ideally we overwrite existing methods (not sure if this works)
```suggestion
        def __init__(self, stream_or_string, **options):
            if isinstance(stream_or_string, bytes):
                stream_or_string = stream_or_string.decode()
            try:
                objects = csv.DictReader(stream_or_string)
            except Exception as exc:
                raise DeserializationError() from exc
            super().__init__(objects, **options)

        def _handle_object(self, obj):
            try:
                model_fields = apps.get_model(obj["model"])._meta.fields
                obj["fields"] = {
                    field.name: obj[field.name]
                    for field in model_fields
                    if field.name in obj
                }
                yield from super()._handle_object(obj)
            except (GeneratorExit, DeserializationError):
                raise
            except Exception as exc:
               raise DeserializationError(f"Error deserializing object: {exc}") from exc
```

**Comment by AMK9978** [comment]
@sarahboyce 
Spot on! And sorry for the inconvenience. I should've tweaked my code before pushing! 
I just added
```
if isinstance(stream_or_string, str):

                stream_or_string = stream_or_string.splitlines()
```
to the deserializer to enable it to work with `str` input, along with file input and bytes.
Btw, I was thinking of adding some tests to cover this line of json deserializer:
```
        if not isinstance(stream_or_string, (bytes, str)):
            stream_or_string = stream_or_string.read()
```
I don't know when Django wants to support type declaration but for now input types to the deserializers don't seem crystal clear (Not a big deal).

**Comment by sarahboyce** [comment]
Just tested serializing and deserializing our Joe and Jane with the csv serializer and it worked for me :tada: 

> I don't know when Django wants to support type declaration but for now input types to the deserializers don't seem crystal clear (Not a big deal).

I don't think types will be added any time soon (it's quite a topic)

**Review by sarahboyce — APPROVED** [review]
Thank you for your work on this @AMK9978 :star:

**Pull request merged**
